File: CourceWork2Java/src/Library.java
import java.time.LocalDate;
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
//import java.util.Comparator;
//import java.util.HashMap;
import java.util.List;
import java.util.Optional;
import java.util.Scanner;

public class Library {
	private List<Book> books = new ArrayList<>();
    private List<User> users = new ArrayList<>();
    private List<Transaction> transactions = new ArrayList<>();
    private List<Account> accounts = new ArrayList<>();
    
    private SearchEngine searchEngine = new SearchEngine();
    private ReportGenerator reportGenerator = new ReportGenerator();
    private FineCalculator fineCalculator = new FineCalculator();
    
    private Scanner scanner;
    
    private int nextBookId = 1;
    private int nextUserId = 1;
    private int nextTransactionId = 1;

    private static final String BOOK_FILE = "books.txt";
    private static final String USER_FILE = "users.txt";
    private static final String TRANSACTION_FILE = "transactions.txt";
    private static final String ACCOUNT_FILE = "accounts.txt";

    // --- Session Management ---
    private String activeUser = null;
    private Role activeRole = null;
    private int activeUserId = 0;

    public Library() {
    	this.scanner = new Scanner(System.in);
        loadAllData();
    }
    
    private <T extends Serializable> List<T> loadData(String filename) {
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(filename))) {
            return (List<T>) ois.readObject();
        } catch (FileNotFoundException e) {
            System.out.println("Файл данных " + filename + " не найден. Начинаем с нуля.");
            return new ArrayList<>();
        } catch (IOException | ClassNotFoundException e) {
            System.err.println("Ошибка при загрузке данных из " + filename + ": " + e.getMessage());
            return new ArrayList<>();
        }
    }

    private void saveData(String filename, List<?> data) {
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(filename))) {
            oos.writeObject(data);
        } catch (IOException e) {
            System.err.println("Ошибка при сохранении данных в " + filename + ": " + e.getMessage());
        }
    }
    
    private void loadAllData() {
        this.books = loadData(BOOK_FILE);
        this.users = loadData(USER_FILE);
        this.transactions = loadData(TRANSACTION_FILE);
        this.accounts = loadData(ACCOUNT_FILE);
        
        // Инициализация следующих ID (упрощено: берем max + 1)
        // В реальной системе это должно быть надежнее
        if (!books.isEmpty()) nextBookId = books.stream().mapToInt(Book::getId).max().getAsInt() + 1;
        if (!users.isEmpty()) nextUserId = users.stream().mapToInt(User::getId).max().getAsInt() + 1;
        if (!transactions.isEmpty()) nextTransactionId = transactions.stream().mapToInt(Transaction::getId).max().getAsInt() + 1;
    }
    
    public void saveAllData() {
        saveData(BOOK_FILE, books);
        saveData(USER_FILE, users);
        saveData(TRANSACTION_FILE, transactions);
        saveData(ACCOUNT_FILE, accounts);
        System.out.println("\n[СИСТЕМА] Данные сохранены.");
    }

    // Public methods
//    public void displayLibraryInfo() {
//        System.out.println("Library Name: " + name +
//                           ", Address: " + address);
//        System.out.println("Current books: " + books.size());
//        System.out.println("Current users: " + users.size());
//        System.out.println("Current transactions: " + transactions.size());
//    }

    // Methods for working with books
 // --- Вспомогательные методы поиска ---
    private Book findBook(int id) {
        return books.stream().filter(b -> b.getId() == id).findFirst().orElse(null);
    }
    private User findUser(int id) {
        return users.stream().filter(u -> u.getId() == id).findFirst().orElse(null);
    }
    private Account findAccount(String username) {
        return accounts.stream().filter(a -> a.getUsername().equals(username)).findFirst().orElse(null);
    }
    
    // --- Аутентификация ---

    public boolean registerUser() {
        System.out.println("--- РЕГИСТРАЦИЯ (Роль по умолчанию: USER) ---");
        System.out.print("Введите желаемый логин: "); 
        String username = scanner.nextLine();
        
        if (findAccount(username) != null) {
            System.out.println("Ошибка: Логин уже занят.");
            return false;
        }

        System.out.print("Введите пароль: "); 
        String password = scanner.nextLine();

        users.add(new User(nextUserId++, username, "N/A"));
        int newUserId = users.get(users.size() - 1).getId();

        accounts.add(new Account(username, password, Role.USER, newUserId));
        
        System.out.println("Регистрация успешна! Ваш User ID: " + newUserId);
        return true;
    }
    
    public boolean login(String username, String password) {
        Account acc = findAccount(username);

        if (acc == null || !acc.getPassword().equals(password)) {
            System.out.println("Ошибка: Логин или пароль неверны.");
            return false;
        }

        this.activeUser = acc.getUsername();
        this.activeRole = acc.getRole();
        this.activeUserId = acc.getAssociatedId();
        
        System.out.println("Вход успешен. Роль: " + this.activeRole);
        return true;
    }
    
    // Геттеры для сессии
    public Role getActiveRole() { return activeRole; }
    public String getActiveUser() { return activeUser; }
    public int getActiveUserId() { return activeUserId; }

    // --- CRUD / Операции ---

    public void addBook(boolean isDigital) {
    	System.out.println("\n--- ДОБАВЛЕНИЕ НОВОЙ КНИГИ ---");
        System.out.print("Введите название: "); String title = scanner.nextLine();
        System.out.print("Введите автора: "); String author = scanner.nextLine();
        System.out.print("Введите ISBN: "); String isbn = scanner.nextLine();
        
        String path = "N/A";
        if (isDigital) {
            System.out.print("Введите полный путь к файлу книги (например, C:\\\\eBooks\\\\book.pdf): "); 
            path = scanner.nextLine();
        }

        books.add(new Book(nextBookId++, title, author, isbn, path));
        System.out.println("Книга добавлена с ID: " + books.get(books.size() - 1).getId());
    }
    
    public void addUser() {
        System.out.println("\n--- ДОБАВЛЕНИЕ НОВОГО ПОЛЬЗОВАТЕЛЯ ---");
        System.out.print("Введите имя пользователя: ");
        String name = scanner.nextLine();
        
        // Контакт задается вручную или "N/A"
        System.out.print("Введите контактную информацию (email/телефон, или Enter для пропуска): ");
        String contact = scanner.nextLine();
        if (contact.trim().isEmpty()) {
            contact = "N/A";
        }

        users.add(new User(nextUserId++, name, contact));
        System.out.println("Пользователь успешно добавлен. User ID: " + users.get(users.size() - 1).getId());
    }
    
    public void updateBook() {
        System.out.println("\n--- ОБНОВЛЕНИЕ ИНФОРМАЦИИ О КНИГЕ ---");
        System.out.print("Введите ID книги для обновления: ");
        
        try {
            int id = scanner.nextInt();
            scanner.nextLine(); // Consume newline

            Book bookToUpdate = findBook(id);
            if (bookToUpdate == null) {
                System.out.println("Ошибка: Книга с ID " + id + " не найдена.");
                return;
            }

            System.out.println("Текущее название: " + bookToUpdate.getTitle());
            System.out.print("Новое название (Enter для пропуска): "); 
            String newTitle = scanner.nextLine();
            if (!newTitle.isEmpty()) {
                System.out.println("Название обновлено.");
            }
            
            System.out.println("Текущий автор: " + bookToUpdate.getAuthor());
            System.out.print("Новый автор (Enter для пропуска): "); 
            String newAuthor = scanner.nextLine();
            if (!newAuthor.isEmpty()) {
                // bookToUpdate.setAuthor(newAuthor);
                System.out.println("Автор обновлен.");
            }
            
            System.out.println("Текущий ISBN: " + bookToUpdate.getIsbn());
            System.out.print("Новый ISBN (Enter для пропуска): "); 
            String newIsbn = scanner.nextLine();
            if (!newIsbn.isEmpty()) {
                // bookToUpdate.setIsbn(newIsbn);
                System.out.println("ISBN обновлен.");
            }

            System.out.println("Информация о книге ID " + id + " обновлена.");

        } catch (Exception e) {
            System.out.println("Некорректный ввод ID.");
            scanner.nextLine();
        }
    }
    
    public void borrowBook() {
        System.out.print("Введите ID пользователя: "); int uId = scanner.nextInt(); scanner.nextLine();
        System.out.print("Введите ID книги для выдачи: "); int bId = scanner.nextInt(); scanner.nextLine();
        
        User user = findUser(uId);
        Book book = findBook(bId);

        if (user == null) { System.out.println("Ошибка: Пользователь не найден."); return; }
        if (book == null || !book.getAvailable()) { System.out.println("Ошибка: Книга не найдена или выдана."); return; }

        transactions.add(new Transaction(nextTransactionId++, bId, uId));
        book.setAvailability(false);
        System.out.println("Книга '" + book.getTitle() + "' выдана.");
    }

    public void returnBook() {
        System.out.print("Введите ID книги, которую возвращают: "); int bId = scanner.nextInt(); scanner.nextLine();

        Book book = findBook(bId);
        if (book == null) { System.out.println("Ошибка: Книга не найдена."); return; }
        
        Transaction activeT = transactions.stream()
            .filter(t -> t.getBookId() == bId && t.isActiveStatus())
            .findFirst().orElse(null);

        if (activeT == null) { System.out.println("Ошибка: Активная запись о выдаче не найдена."); return; }

        book.setAvailability(true);
        activeT.completeTransaction();
        
        double fine = fineCalculator.calculateFine(activeT);
        System.out.println("Книга '" + book.getTitle() + "' успешно возвращена.");
        if (fine > 0) {
            System.out.println("!!! Начислен штраф: " + fine + " у.е. за просрочку.");
        }
    }

    public void viewMyBooks() {
        System.out.println("\n--- МОИ ВЫДАННЫЕ КНИГИ ---");
        boolean found = false;
        
        for (Transaction t : transactions) {
            if (t.getUserId() == activeUserId && t.isActiveStatus()) {
                found = true;
                Book book = findBook(t.getBookId());
                
                if (book != null) {
                    System.out.print("[" + book.getId() + "] " + book.getTitle() 
                         + " (Срок: " + t.getDueDate() + ")");
                    
                    if (!book.getFilePath().equals("N/A")) {
                        System.out.print(" [ДОСТУПНО ЧТЕНИЕ]");
                    }
                    System.out.println();
                }
            }
        }
        if (!found) {
            System.out.println("У вас нет активных выданных книг.");
            return;
        }

        System.out.print("\nВведите ID книги, которую хотите прочитать (или 0 для выхода): "); 
        int choiceId = scanner.nextInt(); scanner.nextLine();

        if (choiceId != 0) {
            Book targetBook = findBook(choiceId);
            if (targetBook != null && !targetBook.getFilePath().equals("N/A")) {
                // Эмуляция открытия файла
                SystemHelper.openFile(targetBook.getFilePath());
            } else {
                System.out.println("Книга не найдена или цифровая копия недоступна.");
            }
        }
    }
    
    public void displayAllBooks() {
        System.out.println("\n--- СПИСОК ВСЕХ КНИГ ---");
        if (books.isEmpty()) {
            System.out.println("В библиотеке нет книг.");
            return;
        }
        books.forEach(Book::display);
    }
    
    public void displayAllUsers() {
        System.out.println("\n--- СПИСОК ВСЕХ ПОЛЬЗОВАТЕЛЕЙ ---");
        if (users.isEmpty()) {
            System.out.println("Пользователи не зарегистрированы.");
            return;
        }
        users.forEach(User::display);
    }
    
    public void displayAllTransactions() {
        System.out.println("\n--- ВСЕ ТРАНЗАКЦИИ ---");
        if (transactions.isEmpty()) {
            System.out.println("Транзакции отсутствуют.");
            return;
        }
        transactions.forEach(Transaction::displayInfo);
    }
    
    public void searchMenu() {
        System.out.println("\n--- МЕНЮ ПОИСКА ---");
        System.out.println("1. Поиск по названию");
        System.out.println("2. Поиск по автору");
        System.out.print("Ваш выбор: ");
        
        try {
            int choice = scanner.nextInt();
            scanner.nextLine(); // Consume newline
            
            System.out.print("Введите поисковый запрос: ");
            String query = scanner.nextLine();
            
            List<Book> results = new ArrayList<>();

            switch (choice) {
                case 1:
                    results = searchEngine.searchByTitle(books, query);
                    break;
                case 2:
                    results = searchEngine.searchByAuthor(books, query);
                    break;
                default:
                    System.out.println("Неверный выбор.");
                    return;
            }
            
            System.out.println("\n--- РЕЗУЛЬТАТЫ ПОИСКА (" + results.size() + ") ---");
            if (results.isEmpty()) {
                System.out.println("Ничего не найдено.");
            } else {
                results.forEach(Book::display);
            }

        } catch (Exception e) {
            System.out.println("Некорректный ввод.");
            scanner.nextLine(); 
        }
    }
    
    public void generateReports() {
        System.out.println("\n--- ГЕНЕРАЦИЯ ОТЧЕТОВ ---");
        System.out.println("1. Отчет об активных выдачах");
        System.out.println("2. Отчет о штрафах (по активным выданным)");
        System.out.print("Ваш выбор: ");

        try {
            int choice = scanner.nextInt();
            scanner.nextLine(); 

            switch (choice) {
                case 1:
                    reportGenerator.generateActiveLoansReport(transactions, users, books);
                    break;
                case 2:
                    generateFineReport();
                    break;
                default:
                    System.out.println("Неверный выбор.");
            }
        } catch (Exception e) {
            System.out.println("Некорректный ввод.");
            scanner.nextLine(); 
        }
    }

    private void generateFineReport() {
        System.out.println("\n--- ОТЧЕТ О ШТРАФАХ ---");
        boolean foundFine = false;
        
        for (Transaction t : transactions) {
            if (t.isActiveStatus()) {
                double fine = fineCalculator.calculateFine(t);
                if (fine > 0) {
                    foundFine = true;
                    
                    User u = findUser(t.getUserId());
                    Book b = findBook(t.getBookId());
                    
                    String userName = (u != null) ? u.getName() : "ID " + t.getUserId() + " (Не найден)";
                    String bookTitle = (b != null) ? b.getTitle() : "ID " + t.getBookId() + " (Не найден)";
                    
                    System.out.printf("Пользователь: %s | Книга: %s | Штраф: %.2f\n", userName, bookTitle, fine);
                }
            }
        }
        
        if (!foundFine) {
            System.out.println("На данный момент штрафов нет.");
        }
    }
}

File: Library.cpp
#define _CRT_SECURE_NO_WARNINGS
#include "Library.hpp"
#include "Book.hpp"
#include "User.hpp"
#include "Transaction.hpp"
#include "FineCalculator.hpp"
#include "SearchEngine.hpp"
#include "ReportGenerator.hpp"
#include <fstream>
#include <algorithm>

using namespace std;

// constructors and destructor
Library::Library() : nextBookId(1), nextUserId(1), nextTransactionId(1) {
    loadBooks();
    loadUsers();
    loadTransactions();
    loadAccounts();
}

Library::~Library() {
    saveAllData();
}

// --- I/O implementations ---
void Library::saveBooks() {
    ofstream ofs(BOOK_FILE);
    for (const auto& book : books) { ofs << book.serialize() << endl; }
    ofs.close();
}

void Library::loadBooks() {
    ifstream ifs(BOOK_FILE);
    string line;
    books.clear();
    int maxId = 0;
    while (getline(ifs, line)) {
        Book book;
        book.deserialize(line);
        books.push_back(book);
        if (book.getId() > maxId) maxId = book.getId();
    }
    nextBookId = maxId + 1;
    ifs.close();
}

void Library::saveUsers() {
    ofstream ofs(USER_FILE);
    for (const auto& user : users) { ofs << user.serialize() << endl; }
    ofs.close();
}

void Library::loadUsers() {
    ifstream ifs(USER_FILE);
    string line;
    users.clear();
    int maxId = 0;
    while (getline(ifs, line)) {
        User user;
        user.deserialize(line);
        users.push_back(user);
        if (user.getId() > maxId) maxId = user.getId();
    }
    nextUserId = maxId + 1;
    ifs.close();
}

void Library::saveTransactions() {
    ofstream ofs(TRANSACTION_FILE);
    for (const auto& t : transactions) { ofs << t.serialize() << endl; }
    ofs.close();
}

void Library::loadTransactions() {
    ifstream ifs(TRANSACTION_FILE);
    string line;
    transactions.clear();
    int maxId = 0;
    while (getline(ifs, line)) {
        Transaction t;
        t.deserialize(line);
        transactions.push_back(t);
        if (t.getTransactionId() > maxId) maxId = t.getTransactionId();
    }
    nextTransactionId = maxId + 1;
    ifs.close();
}

void Library::saveAccounts() {
    ofstream ofs(ACCOUNT_FILE);
    for (const auto& acc : accounts) { ofs << acc.serialize() << endl; }
    ofs.close();
}

void Library::loadAccounts() {
    ifstream ifs(ACCOUNT_FILE);
    string line;
    accounts.clear();
    while (getline(ifs, line)) {
        Account acc;
        acc.deserialize(line);
        accounts.push_back(acc);
    }
    ifs.close();
}

void Library::saveAllData() {
    saveBooks();
    saveUsers();
    saveTransactions();
    saveAccounts();
    cout << "\n[СИСТЕМА] Данные сохранены." << endl;
}

// --- Вспомогательные функции ---
Book* Library::findBook(int id) {
    for (auto& book : books) {
        if (book.getId() == id) return &book;
    }
    return nullptr;
}

User* Library::findUser(int id) {
    for (auto& user : users) {
        if (user.getId() == id) return &user;
    }
    return nullptr;
}

Account* Library::findAccount(const std::string& username) {
    for (auto& acc : accounts) {
        if (acc.getUsername() == username) return &acc;
    }
    return nullptr;
}

// --- Methods for working with books ---
void Library::addBook(bool is_digital)
{
    string title, author, isbn;
    cout << "Введите название: "; getline(cin, title);
    cout << "Введите автора: "; getline(cin, author);
    cout << "Введите ISBN: "; getline(cin, isbn);

    string path = "N/A";
    if (is_digital) {
        cout << "Введите полный путь к файлу книги (например, C:\\books\\Name_book.pdf): ";
        getline(cin, path);
    }

    books.emplace_back(nextBookId++, title, author, isbn);
    books.back().setFilePath(path); // Setting the path
    cout << "Книга добавлена с ID: " << books.back().getId() << endl;
}

void Library::viewMyBooks(int userId) const {
    cout << "\n--- МОИ ВЫДАННЫЕ КНИГИ ---" << endl;
    bool found = false;

    for (const auto& t : transactions) {
        if (t.getUserId() == userId && t.isActiveStatus()) {
            found = true;
            const Book* book = nullptr;
            for (const auto& b : books) {
                if (b.getId() == t.getBookId()) {
                    book = &b;
                    break;
                }
            }

            if (book) {
                cout << "[" << book->getId() << "] " << book->getTitle()
                    << " (Срок: " << t.getDueDate() << ")";

                if (book->getFilePath() != "N/A") {
                    cout << " [ЦЕНЗУРИРОВАНО - ДОСТУПНО ЧТЕНИЕ]";
                }
                cout << endl;
            }
        }
    }
    if (!found) {
        cout << "У вас нет активных выданных книг." << endl;
        return;
    }

    int choiceId;
    cout << "\nВведите ID книги, которую хотите прочитать (или 0 для выхода): ";
    cin >> choiceId;
    cin.ignore();

    if (choiceId != 0) {
        const Book* targetBook = nullptr;
        for (const auto& b : books) {
            if (b.getId() == choiceId && b.getFilePath() != "N/A") {
                targetBook = &b;
                break;
            }
        }

        if (targetBook) {
            // use an auxiliary class to open the file.
            SystemHelper::openFile(targetBook->getFilePath());
        }
        else {
            cout << "Книга не найдена или цифровая копия недоступна." << endl;
        }
    }
}

void Library::displayAllBooks() const
{
    cout << "\n--- СПИСОК КНИГ ---" << endl;
    if (books.empty()) { cout << "Библиотека пуста." << endl; return; }
    for (const auto& book : books) { book.display(); }
}

void Library::updateBook() {
    int id;
    cout << "Введите ID книги для изменения: ";
    cin >> id;
    cin.ignore();

    Book* book = findBook(id);
    if (!book) {
        cout << "Книга не найдена." << endl;
        return;
    }

    cout << "Текущая книга: "; book->display();

    string newTitle, newAuthor, newIsbn;
    cout << "Новое название (Enter для пропуска): "; getline(cin, newTitle);
    if (!newTitle.empty()) book->title = newTitle;

    cout << "Новый автор (Enter для пропуска): "; getline(cin, newAuthor);
    if (!newAuthor.empty()) book->author = newAuthor;

    cout << "Новый ISBN (Enter для пропуска): "; getline(cin, newIsbn);
    if (!newIsbn.empty()) book->isbn = newIsbn;

    cout << "Книга обновлена." << endl;
}

void Library::borrowBook() {
    int uId, bId;
    cout << "Введите ID пользователя: "; cin >> uId;
    cout << "Введите ID книги для выдачи: "; cin >> bId;
    cin.ignore();

    User* user = findUser(uId);
    Book* book = findBook(bId);

    if (!user) { cout << "Ошибка: Пользователь не найден." << endl; return; }
    if (!book) { cout << "Ошибка: Книга не найдена." << endl; return; }
    if (!book->getAvailability()) { cout << "Ошибка: Книга уже выдана." << endl; return; }

    transactions.emplace_back(nextTransactionId++, bId, uId);
    book->setAvailability(false);

    cout << "Книга '" << book->getTitle() << "' выдана пользователю "
        << user->getName() << ". Срок: " << transactions.back().getDueDate() << endl;
}

void Library::returnBook() {
    int bId;
    cout << "Введите ID книги, которую возвращают: "; cin >> bId;
    cin.ignore();

    Book* book = findBook(bId);
    if (!book) { cout << "Ошибка: Книга не найдена." << endl; return; }

    Transaction* activeT = nullptr;
    for (auto& t : transactions) {
        if (t.getBookId() == bId && t.isActiveStatus()) {
            activeT = &t;
            break;
        }
    }

    if (!activeT) { cout << "Ошибка: Активная запись о выдаче для этой книги не найдена." << endl; return; }

    book->setAvailability(true);
    activeT->completeTransaction();

    double fine = fineCalculator.calculateFine(*activeT);
    cout << "Книга '" << book->getTitle() << "' успешно возвращена." << endl;
    if (fine > 0) {
        cout << "!!! Начислен штраф: " << fine << " у.е. за просрочку." << endl;
    }
}

// --- Methods for working with users ---
void Library::addUser()
{
    string name, contact;
    cout << "Введите имя пользователя: "; getline(cin, name);
    cout << "Введите контакт (телефон/email): "; getline(cin, contact);

    users.emplace_back(nextUserId++, name, contact);
    cout << "Пользователь добавлен с ID: " << users.back().getId() << endl;
}

void Library::displayAllUsers() const
{
    cout << "\n--- СПИСОК ПОЛЬЗОВАТЕЛЕЙ ---" << endl;
    if (users.empty()) { cout << "Нет зарегистрированных пользователей." << endl; return; }
    for (const auto& user : users) { user.display(); }
}

void Library::displayAllTransactions() const
{
    cout << "\n--- ИСТОРИЯ ТРАНЗАКЦИЙ ---" << endl;
    if (transactions.empty()) { cout << "Транзакций нет." << endl; return; }
    for (const auto& t : transactions) { t.display(); }
}

// --- Authentication ---
bool Library::registerUser() {
    string username, password;
    cout << "--- РЕГИСТРАЦИЯ (Роль по умолчанию: USER) ---" << endl;
    cout << "Введите желаемый логин: "; getline(cin, username);

    if (findAccount(username)) {
        cout << "Ошибка: Логин уже занят." << endl;
        return false;
    }

    cout << "Введите пароль: "; getline(cin, password);

    // registering a new user in the user list
    users.emplace_back(nextUserId++, username, "N/A"); // Contact initially "N/A"
    int newUserId = users.back().getId();

    // Registering an account with the USER role associated with the new ID
    accounts.emplace_back(username, password, Role::USER, newUserId);

    cout << "Регистрация успешна! Ваш User ID: " << newUserId << endl;
    return true;
}

bool Library::login(std::string& currentUsername, Role& currentRole, int& currentUserId) {
    string username, password;
    cout << "--- ВХОД В СИСТЕМУ ---" << endl;
    cout << "Логин: "; getline(cin, username);
    cout << "Пароль: "; getline(cin, password);

    Account* acc = findAccount(username);

    if (!acc) {
        cout << "Ошибка: Логин или пароль неверны." << endl;
        return false;
    }

    if (acc->getPassword() == password) {
        currentUsername = acc->getUsername();
        currentRole = acc->getRole();
        currentUserId = acc->getAssociatedId();
        cout << "Вход успешен. Роль: " << (currentRole == Role::ADMIN ? "АДМИНИСТРАТОР" : "ПОЛЬЗОВАТЕЛЬ") << endl;
        return true;
    }
    else {
        cout << "Ошибка: Логин или пароль неверны." << endl;
        return false;
    }
}


// --- Specialized ---
void Library::generateReports() const
{
    int choice;
    cout << "\n[ОТЧЕТЫ] 1. Активные выдачи | 2. Штрафы (только для активных): ";
    cin >> choice;
    cin.ignore();

    if (choice == 1) {
        reportGenerator.generateActiveLoansReport(transactions, users, books);
    }
    else if (choice == 2) {
        cout << "\n--- ОТЧЕТ О ПОТЕНЦИАЛЬНЫХ ШТРАФАХ ---" << endl;
        bool foundFine = false;
        for (const auto& t : transactions) {
            if (t.isActiveStatus()) {
                double fine = fineCalculator.calculateFine(t);
                if (fine > 0) {
                    foundFine = true;
                    cout << "Транзакция ID: " << t.getTransactionId() << " | Штраф: " << fine << " у.е." << endl;
                }
            }
        }
        if (!foundFine) cout << "Активных просроченных книг нет." << endl;
    }
    else {
        cout << "Неверный выбор." << endl;
    }
}

void Library::searchMenu()
{
    int choice;
    string query;
    cout << "\n[ПОИСК] 1. По названию | 2. По автору: ";
    cin >> choice;
    cin.ignore();

    if (choice == 1) {
        cout << "Введите часть названия: "; getline(cin, query);
        auto results = searchEngine.searchByTitle(books, query);
        if (results.empty()) cout << "Ничего не найдено." << endl;
        for (const auto& b : results) b.display();
    }
    else if (choice == 2) {
        cout << "Введите часть имени автора: "; getline(cin, query);
        auto results = searchEngine.searchByAuthor(books, query);
        if (results.empty()) cout << "Ничего не найдено." << endl;
        for (const auto& b : results) b.display();
    }
    else {
        cout << "Неверный выбор." << endl;
    }
}